File: algorithms/medium/domino_and_tromino_tilling.ex
defmodule Solution do
  @spec num_tilings(n :: integer) :: integer
  def num_tilings(n) do
    dp = %{1 => 1, 2 => 2, 3 => 5}

    4..n//1
    |> Enum.reduce(dp, fn n, dp ->
      v = rem(dp[n - 1] * 2 + dp[n - 3], 1_000_000_007)
      Map.put(dp, n, v)
    end)
    |> Map.get(n)
  end
end

File: algorithms/medium/domino_and_tromino_tilling.cpp
class Solution {
public:
    int numTilings(int n) {
      int a = 1, b = 2, c = 5;
      if (n == 1) return a;
      if (n == 2) return b;
      if (n == 3) return c;
      for (int i = 4; i <= n; i++) {
        int d = (c * 2 % 1000000007 + a) % 1000000007;
        a = b;
        b = c;
        c = d;
      }
      return c;
    }
};
